feat: draw/show SuperComponents in detail, expand it and show it's internal components in the visualisation diagram

davidsbatista opened the pull request:
### Related Issues

- fixes #9299 

### Proposed Changes:

- Added two new methods to the `Pipeline`
   -  ` _find_super_components(self)` - this will find all the SuperComponents in a Pipeline.
   -   `_merge_super_component_pipelines(self)`: this will expand all the SuperComponents in a pipeline. It will adjust the connections such that the internal components of a SuperComponent connect with the existing components of the Pipeline.

### How did you test it?

- unit tests, integration tests, manual verification


### Notes for the reviewer

In case you want to test it, start a local mermaid server

```bash
docker run --platform linux/amd64 --publish 3000:3000 --cap-add=SYS_ADMIN ghcr.io/jihchi/mermaid.ink
```

Generate an image for the original pipeline, and the version with the expanded components

```python
from pathlib import Path

from haystack import Pipeline
from haystack.components.preprocessors import DocumentPreprocessor
from haystack.components.writers import DocumentWriter
from haystack.document_stores.in_memory import InMemoryDocumentStore

document_store = InMemoryDocumentStore()

pipeline = Pipeline()
pipeline.add_component("preprocessor", DocumentPreprocessor())
pipeline.add_component("writer", DocumentWriter(document_store = document_store))
pipeline.connect("preprocessor", "writer")

# original pipeline
out_file = Path("original_pipeline.png")
pipeline.draw(out_file, server_url="http://localhost:3000")

# merged pipeline that includes all components
out_file = Path("merged_pipeline.png")
pipeline.draw(out_file, server_url="http://localhost:3000",  super_component_expansion=True)

# component by component
pp = Pipeline()
pp.add_component("splitter", DocumentSplitter())
pp.add_component("cleaner", DocumentCleaner())
pp.add_component("writer", DocumentWriter(document_store = document_store))

pp.connect("splitter.documents", "cleaner.documents")
pp.connect("cleaner", "writer")

out_file = Path("component_by_component.png")
pipeline.draw(out_file, server_url="http://localhost:3000",  super_component_expansion=True)
```

## SuperComponent
![original_pipeline](https://github.com/user-attachments/assets/28d814e3-85c4-4f90-883c-f2708a08acff)


## SuperComponent expanded (i.e.: super_component_expansion=True)
![merged_pipeline](https://github.com/user-attachments/assets/d3152288-702c-4a82-9e24-c32f8999948e)


## Pipeline with each components defined originaly
![component_by_component](https://github.com/user-attachments/assets/9ad466a7-10de-4fa6-af60-f4d146ab5366)





### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:` and added `!` in case the PR includes breaking changes.
- I documented my code
- I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

sjrl commented:
Hey @davidsbatista could you also add the mermaid picture of your example? I think that would help with the review

sjrl commented:
Thanks @davidsbatista it's looking good! Looking back at the original issue do you think it would be possible to add this dotted box around where the super component is in the expanded view?

davidsbatista commented:
@sjrl did you do that yourself or the community member? Just to understand if that was programmatically or "photoshopped"

sjrl commented:
> @sjrl did you do that yourself or the community member? Just to understand if that was programmatically or "photoshopped"

That was a community member. Originally from this [comment](https://github.com/deepset-ai/haystack-experimental/discussions/189#discussioncomment-12599399).

sjrl commented:
Oh interesting, yeah I think any kind of visual difference would be nice! It would be great though if we could easily distinguish between different super components within the same pipeline. So maybe different shades of the same color? Also would it be possible to add a legend to indicate which color goes with which super component?

davidsbatista commented:
I will try, need to change back the code and play with the data structure so that one can connect a super component name with it's internal components - I might need some extra time for this.

davidsbatista commented:
![merged_pipeline](https://github.com/user-attachments/assets/c525ef3c-d234-4a46-a458-82743510c2a7)

I've added a function to generate variations of a given colour.

This is applied to a pipeline with supercomponent.

vblagoje commented:
> I've added a function to generate variations of a given colour.
> 
> This is applied to a pipeline with supercomponent.

Very nice!!

vblagoje reviewed (dismissed):
Looks nice - would be great to later extract utils related to super components in pipeline and component being a super component!

sjrl commented on haystack/core/pipeline/base.py:
```diff
@@ -669,7 +669,13 @@ def outputs(self, include_components_with_connected_outputs: bool = False) -> Di
+    def show(
+        self,
+        server_url: str = "https://mermaid.ink",
+        super_component_expansion: bool = False,
```
We should add `super_component_expansion` at the bottom of this function otherwise inserting it in the middle is a breaking change.

sjrl commented on haystack/core/pipeline/base.py:
```diff
@@ -704,14 +715,31 @@ def show(self, server_url: str = "https://mermaid.ink", params: Optional[dict] =
+        self,
+        path: Path,
+        server_url: str = "https://mermaid.ink",
+        super_component_expansion: bool = False,
```
Same here we should move the new variable to the bottom of the function

sjrl commented on haystack/core/pipeline/base.py:
```diff
@@ -1175,7 +1222,7 @@ def _write_component_outputs(
             # We either get the value that was produced by the actor or we use the _NO_OUTPUT_PRODUCED class to indicate
             # that the sender did not produce an output for this socket.
-            # This allows us to track if a pre-decessor already ran but did not produce an output.
+            # This allows us to track if a pre-cessor already ran but did not produce an output.
```
typo
```suggestion
            # This allows us to track if a predecessor already ran but did not produce an output.
```

sjrl commented:
Nice the color coding and the legend look great! Just a few other minor comments.

sjrl commented:
> @sjrl I opted for forcing keyword arguments for draw() and show() - I think this is a better solution which also avoids problems if we need to add /remove any other parameters

I agree with you in principle, but enforcing keyword arguments now is also a breaking change right?

julian-risch commented on haystack/core/pipeline/base.py:
```diff
@@ -710,9 +704,42 @@ def show(
         """
+        import warnings
+
+        warnings.warn(
```
Currently, we would show this warning even if there are no positional arguments. Can we show the warning only if there are any positional arguments? For example making use of [inspect.getargvalues ](https://docs.python.org/3/library/inspect.html#inspect.getargvalues) or with something similar to [this](https://stackoverflow.com/a/68432663)?

davidsbatista replied:
I would suggest to show it either way - just calling the function should be a enough to trigger the warning to the user that it will have some breaking  changes in the future.

I don't think it's noisy if we show it even if he/she doesn't call it with positional arguments.

davidsbatista replied:
I don't think it's good to shadow this from the user, especially if the user calls the function for whatever use he/she might have.

Anyway, I will code this only to trigger the warning in case the user calls the draw() or show() function using positional arguments.

davidsbatista commented:
I've added a function decorator to check if the `show()` or `draw()` are ever called with positional arguments, and only in those cases a `DeprecrationWarning` is triggered - I've also added tests for all possible cases for the decorator function.

I added tests to calls to draw() and show() with positional arguments, without positional arguments to check if the warning is triggered and not triggered in the correct cases.

julian-risch reviewed (approved):
Looks good to me! 👍 Probably we will reuse the `args_deprecated` in future when we switch to kwargs only in other places too.
